module/zfs/dsl_crypt.c

```diff
@@ -266,6 +266,38 @@ spa_crypto_key_compare(const void *a, const void *b)
 	return (0);
 }
 
+int
+dmu_objset_crypto_key_compare(objset_t *osa, objset_t *osb)
+{
+	dsl_crypto_key_t *dcka = NULL;
+	dsl_crypto_key_t *dckb = NULL;
+	int ret;
+	uint64_t obja, objb;
+	spa_t *spa;
+
+	spa = dmu_objset_spa(osa);
+	if (spa != dmu_objset_spa(osb))
+		return (1);
+	obja = dmu_objset_ds(osa)->ds_object;
+	objb = dmu_objset_ds(osb)->ds_object;
+
+	ret = spa_keystore_lookup_key(spa, obja, FTAG, &dcka);
+	if (ret != 0)
+		return (ret);
+	ret = spa_keystore_lookup_key(spa, objb, FTAG, &dckb);
+	if (ret != 0) {
+		spa_keystore_dsl_key_rele(spa, dcka, FTAG);
+		return (ret);
+	}
+
+	ret = spa_crypto_key_compare(dcka, dckb);
```

Comment:
This checks that the datasets have the same `dsl_crypto_key`, which includes the wrapping key. I think that means two datasets with the same master key can have different `dsl_crypto_key` if one of them had its encryption root changed.

We could instead compare [`zk_guid`](https://github.com/openzfs/zfs/blob/zfs-2.2.0/include/sys/zio_crypt.h#L83-L84) which should match as long as the underlying master key is the same. I have a branch with this check, I'll clean it up and share. We should also add a test around this.

Reply:
6001dad106a402ca886bde866a3ab07eb342ca10 switches the check to `zk_guid` and also cleans up the comparison logic according to other review comments.

I'm working on test cases. So far, I have found that performing a raw send into the same pool and then changing the key of one of the datasets is a reasonable way to end up with two different datasets with different encryption roots and the same master key.

Reply:
Oh than the ZAP-ID could be different. Maybe you are right. Than spa_crypto_key_compare wouldn't return the truth, if you are right? Not sure, going to debug it in the next days. Thanks for the implementation and the commit.

Reply:
The on-disk structure is the wrapping key, not the master key. It's possible in principle for the same master key to be stored in multiple wrapping keys. In those cases, we would expect `dck_obj` to differ but the master key (and `zk_guid`) to match.

In testing, I verified with the `send | recv`+`change-key` scenario that using `spa_crypto_key_compare` does not allow block cloning, while it works using the commit I linked above.

Reply:
While building tests for this, I ran into a bug where ZFS crashes after `send | recv`+`change-key`+`cp` from another encrypted dataset. Crash log is [here](https://gist.github.com/Majiir/43cd951522b308904154fd4b79c45929).

I still need to get the ZFS test suite running, but for now I have a reproducer in the form of a NixOS VM test. You don't need NixOS to run it, just the Nix package manager, which can run on your (Linux) distro of choice. To run it:

```
nix run github:Majiir/nixpkgs/35f0c1ad00cf4d72eeb7ffcc78e2e2a1037c6a07#nixosTests.zfs-block-cloning.stable.driver
```

You can find the test script [here](https://github.com/Majiir/nixpkgs/commit/35f0c1ad00cf4d72eeb7ffcc78e2e2a1037c6a07#diff-949d799ebb2ee8dffc34c77edbe3cfd589a13867e125b58ad85633c5948007f1R44-R75).

I tried applying https://github.com/openzfs/zfs/pull/15543 and it didn't fix the issue.

Reply:
@Majiir do you tested if the wrapping key is different when using send | recv + something....?
If this is so I would a hint to the original crypto_compare_function used a few minutes ago.

How did you crash this? Are you waiting for the sync before doing anything? Do you think it's related to this PR? Was sick last weekend and didn't had time to test/read your nix.

Reply:
The crash no longer occurs on latest `master` (8ad73bf4497e8c8d71ad3aa93459b70c6a2050c0 at time of writing) and block cloning works between the original filesystem and the send/recv'd one. :tada: